PullRequest for Issue23: AMDS sgm project. Add Amptek support

XuetaoLiu-S opened the pull request:
- reogranized the project files for common usage
- added AMDS_sgm project
- Add Amptek support for AMDS

XuetaoLiu-S committed:
Reimplement AMDS SGM project. which is the the same as the sample project now

XuetaoLiu-S committed:
Import the codes from the current Amptek application

XuetaoLiu-S committed:
refine the codes for AmptekServer

XuetaoLiu-S committed:
Code refactored for Amptek server group so that it is clearly mapped

XuetaoLiu-S committed:
Code refactored for Amptek server group so that it is clearly mapped

XuetaoLiu-S committed:
fix broken build of AMDS sgm

XuetaoLiu-S committed:
rename AMDSThreadedBufferGroup to AMDSBufferGroupManager and make the corresponding changes

XuetaoLiu-S committed:
release resources for destructors

XuetaoLiu-S committed:
Add append function for AMDSBufferGroupManager to append data to the
corresponding bufferGroup

XuetaoLiu-S committed:
Some code refactor

XuetaoLiu-S committed:
Code refactor of AMDSThreadedTCPDataServer, which will encrypt the
signal/slot of AMDSTCPDataServer, so that the client won't worry too
much about the real intance to handle the signal/slots

XuetaoLiu-S committed:
Add all codes of current Amptek project into AMDS project

XuetaoLiu-S committed:
Small code refactor for AMDSCentralServer

XuetaoLiu-S committed:
finishing merging codes

XuetaoLiu-S committed:
add new SpecturalDataHolder

XuetaoLiu-S committed:
Finish code refactor for AmptekSDD123Detector

XuetaoLiu-S committed:
remove some files

XuetaoLiu-S committed:
Fix possible broken build for AmptekSDD123Detector refactor

XuetaoLiu-S committed:
Fix possible broken build for AmptekSDD123Detector refactor

XuetaoLiu-S committed:
Finished the code logic for AmptekSDD123DetectorManager

XuetaoLiu-S committed:
finish the codes with AmptekSDD123EpicsDetectorManager

XuetaoLiu-S committed:
code refactor: remove unused codes and add comments

XuetaoLiu-S committed:
code refactor: remove unused codes and add comments

XuetaoLiu-S committed:
code refactor: remove unused codes and add comments

XuetaoLiu-S committed:
rename AMDSStatusData to AMDSDwellStatusData

XuetaoLiu-S committed:
fix broken build for AMDS_sample.pro and TCPClient.pro

XuetaoLiu-S committed:
fix invalid signal/slot connections

XuetaoLiu-S committed:
small code refactor to remove source/ in the include files

XuetaoLiu-S committed:
remove some commented off codes

XuetaoLiu-S commented on source/Detector/Amptek/AmptekEventDefinitions.cpp:
```diff
@@ -0,0 +1,193 @@
+		pileUpRejection_ = configurationValue;
+	else if(configuration.contains("SCAI"))
+		scaIndices_.append(configurationValue.toInt());
+//TODO: figure out the definitions of these two configurations...
```
When the correct configuration strings are figured out, this will removed and the commented codes will be uncommented

XuetaoLiu-S commented on source/Detector/Amptek/AmptekEventDefinitions.cpp:
```diff
@@ -0,0 +1,193 @@
+	for(int x = 0, size = configurationList.count(); x < size; x++){
+		QString configuration = configurationList.at(x);
+		QString configurationValue = configuration.section('=', -1);
+// TODO: after figuring out the defintions of configurationValueHightIndices / configurationValueLowIndices, we can use this function
```
When the correct configuration strings are figured out, this will removed and the commented codes will be uncommented.

XuetaoLiu-S committed:
add commented off codes in readStatusData() back

XuetaoLiu-S committed:
move the common main function to AMDSMain_common.h

XuetaoLiu-S committed:
change camle case file names to snake case

XuetaoLiu-S committed:
code refactor to make it clearer

XuetaoLiu-S commented on source/DataElement/AMDSThreadedBufferGroup.cpp:
```diff
@@ -1,27 +1,89 @@
+	bufferGroup_->append(value);
+
+	if (bufferGroup_->cumulativeEnabled()) {
+		AMDSDwellSpectralDataHolder *cumulativeDataHolder = qobject_cast<AMDSDwellSpectralDataHolder *>(bufferGroup_->cumulativeDataHolder());
```
This might need to be redesigned in the future to have more general support.

XuetaoLiu-S committed:
Code refactor, and remove unused files

XuetaoLiu-S committed:
move AMDSDwellSpectralDataHolder to AMDSSpectralDataHolder.h

XuetaoLiu-S committed:
remove AmptekSDD123Application.cpp/.h

XuetaoLiu-S commented on source/Detector/Amptek/SGM/AmptekCommandManagerSGM.h:
```diff
@@ -0,0 +1,102 @@
+
+	virtual ~AmptekCommandManagerSGM();
+
+//	QString requestStatusPacketHex() const;
```
These commented codes/functions will be deleted before merge.

I created a function `requestDataPacket()` to replace all those functions. I keep those replaced codes here for the convenience of reviewers to understand the functionality of `requestDataPacket()`.

XuetaoLiu-S commented on source/Detector/Amptek/SGM/AmptekCommandManagerSGM.cpp:
```diff
@@ -0,0 +1,218 @@
+{
+}
+
+//QString AmptekCommandManagerSGM::requestStatusPacketHex() const{
```
These commented codes/functions will be deleted before merge.

I created a function `requestDataPacket()` to replace all those functions. I keep those replaced codes here for the convenience of reviewers to understand the functionality of `requestDataPacket()`.

XuetaoLiu-S commented on source/Detector/Amptek/AmptekSDD123Server.cpp:
```diff
@@ -0,0 +1,570 @@
+	}
+}
+
+//void AmptekSDD123Server::requestStatus(){
```
These commented codes/functions will be deleted before merge.

I created a function `requestDataPacket()` to replace all those functions. I keep those replaced codes here for the convenience of reviewers to understand the functionality of `requestDataPacket()`.

XuetaoLiu-S commented on source/Detector/Amptek/AmptekSDD123Server.h:
```diff
@@ -0,0 +1,133 @@
+	int forwardDatagram(const QByteArray &datagram);
+
+	void requestDataPacket(AmptekCommandManagerSGM::AmptekCommandDefSGM command, const QString &dataStringHex="", bool fakeSendRequest=false, bool sendSyncRequest=false, int overrideTimeout=-1);
+//	void requestStatus();
```
These commented codes/functions will be deleted before merge.

I created a function `requestDataPacket()` to replace all those functions. I keep those replaced codes here for the convenience of reviewers to understand the functionality of `requestDataPacket()`.

XuetaoLiu-S committed:
code refactor

XuetaoLiu-S committed:
make project file more general. Fix broken AMDS_sample project

iainworkman commented on AMDS_sgm.pro:
```diff
@@ -1,7 +1,14 @@
+contains(USERNAME, liux0) {
+	HOME_FOLDER = "/home/liux0"
+}
+
```
Is there a reason why (echo $HOME) doesn't work for your username, and is there likely to be similar issues on any of the other SL machines (which will be the deployment environment)?

XuetaoLiu-S replied:
For new employees, the home folder is `/staff/username` :(

iainworkman commented on TCPClient.pro:
```diff
@@ -4,7 +4,10 @@
 #
 #-------------------------------------------------
 
+PATH_TO_SOURCE = /home/liux0/beamline/programming/AcquamanDataServer/source
```
Will these also need to be generalized for testing on the server? 
(My guess is that although the TCPClient won't be 100% required in order to get the server running down here, and interfacing with Acquaman, it will never-the-less prove a useful diagnostic tool)

iainworkman commented on source/DataElement/AMDSBufferGroup.cpp:
```diff
@@ -1,21 +1,92 @@
+
+	AMDSDataHolder* dataHolderRemoved = dataHolders_.append(newData);
+	if(dataHolderRemoved)
+		delete dataHolderRemoved;
```
Since AMDSDataHolder is a QObject child, this should be:

``` C++
dataHolderRemoved->deleteLater();
```

iainworkman commented on compositCommon/AMDSCommon.pri:
```diff
@@ -4,7 +4,6 @@
 #-------------------------------------------------
 
-
 QT       += core network gui sql
```
Does the DS actually use sql?

XuetaoLiu-S replied:
No. The reason we have `sql` is that we are using some Acquaman codes, such as AMPVControl, which requires `sql` package

iainworkman commented on source/Detector/Amptek/AmptekSDD123DetectorManager.cpp:
```diff
@@ -0,0 +1,307 @@
+#include "AmptekSDD123DetectorManager.h"
+
+#include <QDebug>
```
QDebug can be removed.

iainworkman commented on source/Detector/Amptek/AmptekSDD123Packet.cpp:
```diff
@@ -0,0 +1,268 @@
+#include "AmptekSDD123Packet.h"
+
+#include <QStringBuilder>
+#include <QDebug>
```
QDebug can be removed

iainworkman commented on source/Detector/Amptek/AmptekSDD123Server.cpp:
```diff
@@ -0,0 +1,570 @@
+
+void AmptekSDD123Server::requestDataPacket(AmptekCommandManagerSGM::AmptekCommandDefSGM command, const QString &dataStringHex, bool sendFakeRequest, bool sendSyncRequest, int overrideTimeout)
+{
+	AmptekSDD123Packet requestStatusPacket(packetIDCounter_++, AmptekCommandManagerSGM::sddCommands()->amptekCommand(command).hex(), dataStringHex);
```
This is really a minor style thing, but having an increment in the parameter passing of a function tends to make me edgy. When maintaining I always find some unnecessary thought needs to go into whether the value passed will be the value before increment, or after.
In this case I think:

``` C++
AmptekSDD123Packet requestStatusPacket(packetIDCounter_, AmptekCommandManagerSGM::sddCommands()->amptekCommand(command).hex(), dataStringHex);
packedIDCounter_++;
```

...would be a little more obvious as to the intention of the programmer.

dretrex replied:
I 100% agree with this comment.

iainworkman commented:
Overall this looks really good. 
Obviously there is some commented code which can be removed.
Also I feel like a lot of the Amptek code which was migrated could do with some documentation which is currently absent.
Time Estimate: 3 Hours

XuetaoLiu-S committed:
Code changes based on code review

XuetaoLiu-S committed:
refactor AMDSDwellStatusData to encode the plus and division operations and add documentions

XuetaoLiu-S committed:
add more document comments

XuetaoLiu-S committed:
Fix the bugs:
 - wrapping up initialization
 - bind server with UDP socket

XuetaoLiu-S committed:
- Code refactor AmptekCommands
- fix a command code bug

davidChevrier committed:
Finalize the changes on AmptekEventDefintions

XuetaoLiu-S committed:
fix a bug to parse the configurationdata

XuetaoLiu-S committed:
Fix a silly bug

XuetaoLiu-S committed:
- Fix the bug of forgetting to start the thread for
AmptekDetectorManager
- Fix the bug of forgetting to register the new dwell specutrum
dataholder

iainworkman commented on source/Detector/Amptek/AmptekSDD123Server.cpp:
```diff
@@ -0,0 +1,516 @@
+void AmptekSDD123Server::sendRequestDatagram(AmptekSDD123Packet packet, int overrideTimeout)
+{
+	if(socketLocallyBusy_){
+		qDebug() << "\nPACKET QUEUE IS BUSY (send request datagram)\n";
```
Seeing as there's an error mon already being used here, would it be proper to switch this (and the other packet queue is busy messages) to be error mon info calls?

iainworkman commented on source/Detector/Amptek/AmptekSDD123Server.cpp:
```diff
@@ -0,0 +1,516 @@
+		socketLocallyBusy_ = true;
+		currentRequestPacket_ = packet;
+		currentRequestTime_.restart();
+		//udpDetectorSocket_->writeDatagram(packet.datagram(), packet.datagram().size(), configurationMap_->detectorAddress(), 10001);
```
Is this comment removable?

iainworkman commented on source/Detector/Amptek/AmptekSDD123Server.cpp:
```diff
@@ -0,0 +1,516 @@
+		while(!foundTimedOutPacket && timedOutPackets_.count() > 0){
+			headPacket = timedOutPackets_.dequeue();
+
+//			if(responsePacket.command() == "Comm test - Echo packet Response" && headPacket.command() == "Comm test - Echo packet" && responsePacket.dataString() == headPacket.dataString())
```
There are a number of commented lines in the code below, also. Can these be removed?

iainworkman commented on source/Detector/Amptek/SGM/AmptekCommandManagerSGM.cpp:
```diff
@@ -0,0 +1,213 @@
+	return amptekCommand(ReqeustStatusPacket).hex();
+}
+
+//QString AmptekCommandManagerSGM::requestSpectrumHex() const{
```
Are these (and their definitions in the .h) left commented for a reason, or can they be removed?

iainworkman commented on source/application/AMDSCentralServerSGM.cpp:
```diff
@@ -3,89 +3,146 @@
+	maxBufferSize_ = 20 * 60 * 1000; // 20 minuntes
+
+	configurationMaps_.append(new AmptekSDD123ConfigurationMap("Amptek SDD 239", "amptek:sdd1", QHostAddress("192.168.0.239"), QHostAddress("192.168.0.139"), 12004, QHostAddress("192.168.10.104"), AMDSDataTypeDefinitions::Double, 1024, this));
+//	configurationMaps_.append(new AmptekSDD123ConfigurationMap("Amptek SDD 240", "amptek:sdd2", QHostAddress("192.168.0.240"), QHostAddress("192.168.0.101"), 12004, QHostAddress("192.168.10.104"), AMDSDataTypeDefinitions::Double, 1024, this));
```
Can these commented lines be removed?

iainworkman commented on source/application/AMDSCentralServerSGM.cpp:
```diff
@@ -3,89 +3,146 @@
+	}
 
-	startTimer();
+//	connect(amptekDetectorGroup_, SIGNAL(clearHistrogramData(QString)), this, SLOT(onClearHistrogramData(QString)));
```
Also these re: commented lines

iainworkman commented:
Looks good, just some questions regarding commented lines.
(updated my time est. above)

XuetaoLiu-S committed:
Add helper function to process the responsePacket

XuetaoLiu-S committed:
- change the QTime for AMDSDwellStatus to QDateTime
- set the eventTime of the AMDSDwellSpectrralDataHolder to the
dwellStartTime of the status data

XuetaoLiu-S committed:
Fix broken build with the latest changes

XuetaoLiu-S committed:
refine the printData function

XuetaoLiu-S committed:
revert "Fix broken build with the latest changes"

This reverts commit 23befae52849a1eaf33e7277d5d77c1cbcedac5b.

XuetaoLiu-S committed:
Revert "- change the QTime for AMDSDwellStatus to QDateTime"

This reverts commit fa2da52f80fe78aeff497489022e0624272eafa9.

XuetaoLiu-S committed:
user dwellStartTime as the time of the event time

dretrex commented on source/DataElement/AMDSBufferGroup.h:
```diff
@@ -25,28 +26,31 @@ class AMDSBufferGroup : public QObject
-	/// The total capacity of the buffer
-	inline quint64 maxSize() const;
+	/// returns whether Cumulative is enabled or not
+	inline bool cumulativeEnabled() { return enableCumulative_; }
```
This would be a `const` function.

dretrex commented on source/DataElement/AMDSBufferGroup.h:
```diff
@@ -25,28 +26,31 @@ class AMDSBufferGroup : public QObject
+	AMDSDataHolder *cumulativeDataHolder() const;
+
+	/// returns the dataHolder at given index
+	AMDSDataHolder* at(int index);
```
This too, technically.

dretrex commented on source/DataElement/AMDSBufferGroupInfo.h:
```diff
@@ -49,18 +52,20 @@ class AMDSBufferGroupInfo
-	void setUnits(const QString &units) { units_ = units; }
+	inline void setUnits(const QString &units) { units_ = units; }
+	/// Set the dataType of the bufferGroupInfo
+	inline void setDataType(const AMDSDataTypeDefinitions::DataType dataType) { dataType_ = dataType; }
```
Typically, enums are not `const` unless you were actually wanting a `const int`.

dretrex commented on source/DataElement/AMDSBufferGroupInfo.h:
```diff
@@ -49,18 +52,20 @@ class AMDSBufferGroupInfo
+	inline void setAxes(const QList<AMDSAxisInfo> &axes) { axes_ = axes; }
 	/// Set how the bufferGroup should be flattened
-	void setFlattenMethod(DataFlattenMethod method) { flattenMethod_ = method; }
+	inline void setFlattenMethod(DataFlattenMethod method) { flattenMethod_ = method; }
```
`const DataFlattenMethod &`

XuetaoLiu-S replied:
This is an enum type, which as you commented above, we don't need to be const

dretrex commented on source/DataElement/AMDSDwellStatusData.h:
```diff
@@ -0,0 +1,68 @@
+{
+public:
+	AMDSDwellStatusData(int fastCounts = -1, int slowCounts = -1, double detectorTemperature = -1, double accumulationTime = -1, double liveTime = -1, double realTime = -1, int generalPurposeCounter = -1, const QTime &dwellStartTime = QTime(), const QTime &dwellEndTime = QTime(), const QTime &dwellReplyTime = QTime());
+	AMDSDwellStatusData(const AMDSDwellStatusData &other);
```
No destructor.

dretrex commented on source/DataElement/AMDSFlatArray.cpp:
```diff
@@ -198,3 +199,437 @@ AMDSFlatArray AMDSFlatArray::operator /(const quint32 divisor)
+	return true;
+}
+
+bool AMDSFlatArray::setValue(int index, float   value)
```
Out of curiosity, is there a reason why this uses `float` and the next one uses `double` instead of one with `qreal`?  Since you've gone through all the trouble of using the platform independent qatomic types, it seems weird to me that qreal is not being used too.

XuetaoLiu-S replied:
This is from the legacy codes. @davidChevrier can you answer this? Thanks.

davidChevrier replied:
I recall looking into this at one point. I seem to remember that Qt really guaranteed the meaning of the "qint" and "quint" types but something wasn't guaranteed about "qreal". That's my recollection, but it could be a figment of my imagination.

dretrex commented on source/DataElement/AMDSFlatArray.cpp:
```diff
@@ -198,3 +199,437 @@ AMDSFlatArray AMDSFlatArray::operator /(const quint32 divisor)
+void AMDSFlatArray::clear(){
+	switch(dataType_){
+	case AMDSDataTypeDefinitions::Signed8:
+		vectorQint8_.clear();
```
These could be replaced with a single call of `vectorXYZ_ = QVector<XYZ>()`.

dretrex commented on source/DataElement/AMDSFlatArray.cpp:
```diff
@@ -198,3 +199,437 @@ AMDSFlatArray AMDSFlatArray::operator /(const quint32 divisor)
+	dataType_ = dataType;
+
+	switch(dataType_){
+	case AMDSDataTypeDefinitions::Signed8:
```
After resizing, are the contents of of the array still valid?  Should they be?  If not, you could get the resizing and initializing in one by using the above example (`vectorXYZ_ = QVector<XYZ>(size)`).

XuetaoLiu-S replied:
By far, the usage of this function is only clear and resize. It is never used when it has contents. 

Therefore, under current scenario, it doesn't hurt to change the codes to your suggested way.

dretrex commented on source/DataElement/AMDSThreadedBufferGroup.h:
```diff
@@ -4,27 +4,51 @@
+	Q_OBJECT
 public:
-	AMDSThreadedBufferGroup(AMDSBufferGroup *bufferGroup, QObject *parent = 0);
+	AMDSThreadedBufferGroup(AMDSBufferGroupInfo bufferGroupInfo, quint64 maxCountSize, bool enableCumulative = false, QObject *parent = 0);
```
`const AMDSBufferGroupInfo &`

dretrex commented on source/DataHolder/AMDSGenericFlatArrayDataHolder.h:
```diff
@@ -19,12 +19,14 @@ Q_OBJECT
 	virtual void setData(AMDSFlatArray *inputArray) { inputArray->resetTargetArrayAndReplaceData(&valueFlatArray_); }
 	/// implement the function to return the data string
-	virtual QString printData() { return valueFlatArray_.printData(); }
+	virtual QString printData() { return QString("%1 : %2").arg(eventTime().toString()).arg(valueFlatArray_.printData()); }
```
Const function.

dretrex commented on source/Detector/Amptek/AmptekCommandManager.h:
```diff
@@ -0,0 +1,95 @@
+	AmptekCommand amptekCommandFromHex(const QString &hex) const;
+
+	/// returns the commandId with a given command
+	int commandId(const QString &command);
```
Probably a `const` function?

dretrex commented on source/Detector/Amptek/AmptekEventDefinitions.h:
```diff
@@ -0,0 +1,167 @@
+	AmptekConfigurationData& operator =(const AmptekConfigurationData &other);
+
+	/// helper function to pase the configurations
+	void parseConfiguration(QString configuration);
```
`const QString &`

dretrex commented on source/Detector/Amptek/AmptekEventDefinitions.h:
```diff
@@ -0,0 +1,167 @@
+	AmptekConfigurationValuesEvent() : QEvent( (QEvent::Type)AmptekEventDefinitions::ConfigurationValuesEvent) {}
+
+	/// helper function to pase the list of configurations in the event
+	void parseConfigurations(QStringList configurationList);
```
`const QStringList &`

dretrex commented on source/Detector/Amptek/AmptekSDD123Detector.cpp:
```diff
@@ -0,0 +1,656 @@
+
+		AMDSDwellStatusData statusData(fastCounts(), slowCounts(), detectorTemperature(), accumulationTime(), liveTime(), realTime(), generalPurposeCounter(), event->dwellStartTime_, event->dwellEndTime_, event->dwellReplyTime_);
+
+		AmptekSpectrumEvent *spectrumEvent = new AmptekSpectrumEvent();
```
Since these are events, I'm assuming that the event handler handles the memory management of these once they're done?

XuetaoLiu-S replied:
Yes. the event handler will take care of it

dretrex commented on source/Detector/Amptek/AmptekSDD123Detector.cpp:
```diff
@@ -0,0 +1,656 @@
+	//qDebug() << "Spectrum ready:\n " << spectrum;
+	emit spectrumChanged(spectrum);
+
+	return spectrumArray;
```
Caller is responsible for memory?

XuetaoLiu-S replied:
Yes, the caller added the array to the event object

dretrex commented on source/Detector/Amptek/AmptekSDD123Detector.cpp:
```diff
@@ -0,0 +1,656 @@
+	internalSetBoardTemperature(tmpData.at(0));
+	//qDebug() << "boardTemp: " << tmpData.toHex() << boardTemp;
+
+	tmpData = statusData.mid(35, 1);
```
Question: is the same status data used for multiple things?  You use `mid(35, 1)` 5 times in a row for different things.

XuetaoLiu-S replied:
Yes. This is from the original Ampteck server. We are using bitwise operation here :)

dretrex commented on source/Detector/Amptek/AmptekSDD123Detector.cpp:
```diff
@@ -0,0 +1,656 @@
+
+	//qDebug() << "unusedByte0 " << unusedByte0;
+
+	tmpData = statusData.mid(36, 1);
```
Same with 36.

dretrex commented on source/Detector/Amptek/AmptekSDD123Detector.cpp:
```diff
@@ -0,0 +1,656 @@
+	if(slowCounts != slowCounts_){
+		slowCounts_ = slowCounts;
+		emit slowCountsChanged(slowCounts_);
+		emit slowCountsChanged(QString("%1").arg(slowCounts_));
```
Since this is going to be spitting out a lot of signals, would it make more sense to only emit the `int` versions of the signals and then have the specific slots that listen to it do the string conversion?  I think that doing a lot of string conversion all the time might be a bit wasteful.

XuetaoLiu-S replied:
I agree. It is from the legacy codes though. I will check the usages. If there is no usages, about the QString version signal, I will remove it.

dretrex commented on source/Detector/Amptek/AmptekSDD123Detector.h:
```diff
@@ -0,0 +1,324 @@
+	virtual bool event(QEvent *e);
+
+	/// returns the dataType the detector will return
+	inline AMDSDataTypeDefinitions::DataType dataType() { return dataType_; }
```
`const` function.

dretrex commented on source/Detector/Amptek/AmptekSDD123Detector.h:
```diff
@@ -0,0 +1,324 @@
+	/// returns the dataType the detector will return
+	inline AMDSDataTypeDefinitions::DataType dataType() { return dataType_; }
+	/// returns the buffersize (spectrum size) the detector will return
+	inline int bufferSize() { return bufferSize_; }
```
And here.

dretrex commented on source/Detector/Amptek/AmptekSDD123Packet.h:
```diff
@@ -0,0 +1,146 @@
+	AmptekSDD123Packet(int packetID, const QString &commandHex, const QString &dataString = "", bool isForwarded = false, QObject *parent = 0);
+	AmptekSDD123Packet(int packetID, const QByteArray &datagram, bool isForwarded = false, QObject *parent = 0);
+	AmptekSDD123Packet(const AmptekSDD123Packet &original);
+	AmptekSDD123Packet(int packetID, const AmptekSDD123Packet &original);
```
Destructor?

dretrex commented on source/Detector/Amptek/AmptekSDD123Packet.h:
```diff
@@ -0,0 +1,146 @@
+	/// returns the legth of the data
+	int dataLength() const;
+
+	/// parses and returns the spectrum byte array from the package datagram
```
`const` functions.

dretrex commented on source/Detector/Amptek/AmptekSDD123Server.cpp:
```diff
@@ -0,0 +1,522 @@
+
+		timeOutIDCounter_ = -27;
+
+		requestSpectrumTime_ = new QTime();
```
Times not cleaned up.  Also, there does not appear to be a destructor to destroy them in either.

dretrex commented on source/Detector/Amptek/AmptekSDD123Server.cpp:
```diff
@@ -0,0 +1,522 @@
+	}
+}
+
+QList<double> AmptekSDD123Server::droppedPacketStatistics() const
```
Do we only ever care about the last 30 seconds?  Would it make sense to have a version that took in a value and returned the statistics in that interval?

dretrex commented on source/Detector/Amptek/AmptekSDD123Server.cpp:
```diff
@@ -0,0 +1,522 @@
+
+int AmptekSDD123Server::forwardDatagram(const QByteArray &datagram)
+{
+	AmptekSDD123Packet forwardingPacket(packetIDCounter_++, datagram, true);
```
I would prefer being more explicit here for readability.

``` c++
AmptekSDD123Packet forwardingPacket(packetIDCounter_, datagram, true);
packetIDCounter_++;
```

dretrex commented on source/Detector/Amptek/AmptekSDD123Server.cpp:
```diff
@@ -0,0 +1,522 @@
+
+void AmptekSDD123Server::syncRequest(int timeout)
+{
+	QString hashTag = QString("%1").arg(--timeOutIDCounter_);
```
Readability change here and in fakeSyncRequest.  I know it's being a bit overly pedantic, but I think the ease of reading it in the future will pay off with more than saving a couple lines of code.

dretrex commented on source/Detector/Amptek/AmptekSDD123Server.h:
```diff
@@ -0,0 +1,209 @@
+	/// request all text configuration parameters
+	void requestAllTextConfigurationParameters();
+	/// request text configuration readback with single/multi command request(s)
+	void textConfigurationReadback(QString singleCommandRequest);
```
This header is missing a bunch of `const XYZ &` arguments.

dretrex commented on source/Detector/Amptek/AmptekSDD123Server.h:
```diff
@@ -0,0 +1,209 @@
+
+protected:
+	/// helper function to handle the local timedout message queue for the responsePacket
+	void processLocalStoredPacket(AmptekSDD123Packet responsePacket);
```
These ones in particular are probably very important to have the `const XYZ &` arguments.  Especially the `QByteArray`'s.

dretrex commented on source/Detector/Amptek/AmptekSDD123Server.h:
```diff
@@ -0,0 +1,209 @@
+Q_OBJECT
+
+public:
+	AmptekSDD123Server(AmptekSDD123ConfigurationMap *configurationMap, const QString &serverName, QObject *parent = 0);
```
No destructor.

dretrex commented on source/Detector/Amptek/AmptekSDD123ServerGroup.cpp:
```diff
@@ -0,0 +1,166 @@
+	connect(serverConfigurationStateMapper_, SIGNAL(mapped(int)), this, SLOT(onServerChangedToConfigurationState(int)));
+	connect(serverDwellStateMapper_, SIGNAL(mapped(int)), this, SLOT(onServerChangedToDwellState(int)));
+
+	QTimer::singleShot(1000, this, SLOT(initiateAllRequestSpectrum()));
```
I'm assuming this is a legacy thing, but what is the specific reason for 1 second?

XuetaoLiu-S replied:
My understanding is that we will start to request spectrum after the 1s. Give the system some time to be ready to receive data

dretrex commented on source/Detector/Amptek/AmptekSDD123ServerGroup.cpp:
```diff
@@ -0,0 +1,166 @@
+		if(serversInDwellMode_.contains(amptekServer))
+			serversInDwellMode_.removeAll(amptekServer);
+
+		QTimer::singleShot(masterRequestInterval_+5, amptekServer, SLOT(disableMCAMCS()));
```
What is the reason for the extra time?

XuetaoLiu-S replied:
This is from the legacy codes as well. I don't know the reason as well. @davidChevrier can you answer this? thx

davidChevrier replied:
If I recall correctly, this was intentional. I think I discovered in testing that it was possible to put the server group in its ::ConfigurationState while it still wanted to send back the next packets of data. This meant that calling disableMCAMCS() too soon would leave junk in the buffer and the whole thing would fall apart. Making sure that this was called "a nominally short period of time after the latest time that new data could come back (ie,, masterRequestInterval)" meant that everything worked fine. To be clear, the [dis/en]ableMCAMCS() almost amounts to putting the Amptek into and out of continuous mode. You need to safely take it out of continuous mode to start sending configuration data to it.

dretrex commented on source/Detector/Amptek/AmptekSDD123ServerGroup.h:
```diff
@@ -0,0 +1,109 @@
+	~AmptekSDD123ThreadedDataServerGroup();
+
+	/// returns the AmptekServer at index
+	AmptekSDD123Server* serverAt(int index);
```
`const` function.

dretrex commented on source/Detector/Amptek/AmptekSDD123ServerGroup.h:
```diff
@@ -0,0 +1,109 @@
+	virtual ~AmptekSDD123ServerGroup();
+
+	/// returns the AmptekServer at given index
+	AmptekSDD123Server* serverAt(int index);
```
Here too.

dretrex commented on source/Detector/Amptek/SGM/AmptekCommandManagerSGM.cpp:
```diff
@@ -0,0 +1,213 @@
+	amptekCommands_.append( AmptekCommand(ResponseCommTestEchoPacket,            "8f7f", "Comm test - Echo packet Response"));
+
+	//Request types
+	QStringList tempResponses;
```
These lines would be easier to read as a single line that would do the same thing.

``` c++
QStringList tempResponses = QStringList() << "abc" << "def";
```

dretrex commented on source/Detector/Amptek/SGM/AmptekCommandManagerSGM.h:
```diff
@@ -0,0 +1,100 @@
+	/// implementation of the function to initialize the amptek commands and the hex mappling
+	virtual void initiateAmptekCommands();
+	/// helper function to create response from given texts
+	QStringList internalCreateResponsesFromText(QStringList texts);
```
`const QStringList &`

dretrex commented on source/Detector/Amptek/SGM/AmptekSDD123DetectorGroupSGM.h:
```diff
@@ -0,0 +1,30 @@
+{
+	Q_OBJECT
+public:
+	AmptekSDD123DetectorGroupSGM(QList<AmptekSDD123ConfigurationMap*> configurationMaps, QObject *parent = 0);
```
`const QList<AmptekSDD123ConfigurationMap *> &`

dretrex commented on source/Detector/Amptek/SGM/AmptekSDD123DetectorGroupSGM.h:
```diff
@@ -0,0 +1,30 @@
+	AmptekSDD123DetectorGroupSGM(QList<AmptekSDD123ConfigurationMap*> configurationMaps, QObject *parent = 0);
+	~AmptekSDD123DetectorGroupSGM();
+
+	/// returns the list of AmptekDetectorManager
```
`const` function.

dretrex commented on source/Detector/Amptek/SGM/AmptekSDD123EPICSDetectorManager.cpp:
```diff
@@ -0,0 +1,638 @@
+	connect(slowThresholdControl_, SIGNAL(valueChanged(double)), this, SLOT(onSlowThresholdControlValueChanged(double)));
+	connect(peakingTimeControl_, SIGNAL(valueChanged(double)), this, SLOT(onPeakingTimeControlValueChanged(double)));
+	connect(fastChannelPeakingTimeControl_, SIGNAL(valueChanged(double)), this, SLOT(onFastChannelPeakingTimeControlValueChanged(double)));
+	connect(roi1LowIndexControl_, SIGNAL(valueChanged(double)), this, SLOT(onROI1LowIndexControlValueChanged(double)));
```
This might be a better question for @davidChevrier, but do we care about these ROIs anymore?

dretrex commented on source/Detector/Amptek/SGM/AmptekSDD123EPICSDetectorManager.cpp:
```diff
@@ -0,0 +1,638 @@
+	if(!connected_ || !receivedConfigurationDataOnce_)
+		return;
+	/*
+	if(newValue == 0 && configurationData_.hvSet_ == "OFF")
```
To be implemented later?

XuetaoLiu-S replied:
Maybe... it is commented off in the old Amptek server as well.

dretrex commented on source/Detector/Amptek/SGM/AmptekSDD123EPICSDetectorManager.cpp:
```diff
@@ -0,0 +1,638 @@
+		return;
+	if(roi1LowIndexControl_->withinTolerance(configurationData_.scaLowIndices_.at(0)))
+		return;
+	setDetectorSingleChannelAnalyzer(1, (int)newValue, configurationData_.scaHighIndices_.at(0));
```
I think there is a mismatch here.  It looks like the `scaHighIndices_` are being used in the low index changed and the `scaLowIndices_` are being used in the high index changed slots.

XuetaoLiu-S replied:
@dretrex yes, I think you are right, although the old codes are like this as well. @davidChevrier I believe that this is a bug. I will fix it in AMDS

XuetaoLiu-S replied:
@davidChevrier @dretrex actually, it is NOT. It is supposed to be like this...

XuetaoLiu-S replied:
The prototype of the function `setDetectorSingleChannelAnalyzer` is:
`void setDetectorSingleChannelAnalyzer(int singleChannelAnalyzerIndex, int lowChannelIndex, int highChannelIndex);`

dretrex replied:
Ah, yes.  Good call.

dretrex commented on source/Detector/Amptek/SGM/AmptekSDD123EPICSDetectorManager.h:
```diff
@@ -0,0 +1,132 @@
+
+protected slots:
+	/// slot to handle continuous data update
+	virtual void onContinuousAllDataUpdate(AMDSDataHolder *spectrum, AMDSDwellStatusData statusData, int count, double elapsedTime) ;
```
`const AMDSDwellStatusData &`?

dretrex commented on source/Detector/Amptek/SGM/AmptekSDD123EPICSDetectorManager.h:
```diff
@@ -0,0 +1,132 @@
+	/// slot to handle dwell finish data update
+	virtual void onDwellFinishedAllDataUpdate(AMDSDataHolder *spectrum, AMDSDwellStatusData statusData, int count, double elapsedTime);
+	/// slot to handle configuration value update signal
+	void onConfigurationValuesUpdate(AmptekConfigurationData configurationData);
```
`const AmptekConfigurationData &`

dretrex commented on source/application/AMDSCentralServerSGM.cpp:
```diff
@@ -3,89 +3,146 @@
-	oneScalerDataHolder->setSingleValue(valueAsDouble);
+void AMDSCentralServerSGM::initializeAndStartDataServer()
+{
+	amptekThreadedDataServerGroup_ = new AmptekSDD123ThreadedDataServerGroup(configurationMaps_);
```
I didn't see where this is getting cleaned up.

XuetaoLiu-S replied:
There is a destructor ..

dretrex commented on source/application/AMDSCentralServerSGM.h:
```diff
@@ -2,37 +2,67 @@
+	void onServerChangedToDwellState(int index);
+
+	/// slot to handle clear histrogramData request for a given buffer (detector)
+	void onClearHistrogramData(QString detectorName);
```
A couple `const QString &` missing.

dretrex commented on source/application/AMDSCentralServerSample.cpp:
```diff
@@ -17,69 +17,72 @@ AMDSCentralServerSample::AMDSCentralServerSample(QObject *parent) :
 	oneSpectralDataHolder->setData(&oneSpectralFlatArray);
 
-	amptek1BufferGroup_->append(oneSpectralDataHolder);
+	bufferGroupManagers_.value("Amptek1")->append(oneSpectralDataHolder);
```
Does something deal with the `bufferGroupManagers_` memory?  It appears like the light weight spectral holders are new'ed and added but not cleaned.

XuetaoLiu-S replied:
Yes. `bufferGroupManagers_` is defined in the super class and is also cleared there.

dretrex commented:
Overall, great work @XuetaoLiu-CLS!  I did have some comments and questions that I hope aren't too crazy.
Time Estimate: 2 hours

XuetaoLiu-S committed:
- print out dwellStatusData for the specturalDataHolder

XuetaoLiu-S committed:
Modify the usage of datastream write/read

XuetaoLiu-S committed:
refactor the function read/write of AMDSDwellStatusData to readFromDataStream/writeToDataStream

XuetaoLiu-S committed:
Move the datastream read/write of AMDSAxisInfo from AMDSDataStream to AMDSAxisInfo

XuetaoLiu-S committed:
using amptek sdd 241

XuetaoLiu-S committed:
code refactor: move the read/write of eventData and bufferGroupinfo into the class itself

XuetaoLiu-S committed:
code refactor based on PR comments

XuetaoLiu-S committed:
Fix broken build

XuetaoLiu-S committed:
code refactor: remove the usage of AMDSDataStream

XuetaoLiu-S committed:
code refactor: remove the const qualifier for writeToDataStream functions

davidChevrier committed:
Changed the AMDS_sgm.pro to refer to Sheldon's copy of Acquaman and MPlot. Also changed AmptekSDD123Detector::readSpectrumData call so that the AMDSFlatArray data is filled correctly.

davidChevrier committed:
Merge branch 'Issue23' of https://github.com/acquaman/AcquamanDataServer into Issue23

XuetaoLiu-S committed:
more code clean up

XuetaoLiu-S committed:
remove compile warnings

XuetaoLiu-S committed:
add special project configuration for SGM

davidChevrier committed:
Sheldon made changes to the .pro files so that the pathing is clearer for includes.

XuetaoLiu-S committed:
Solve the issue of NULL receiver of postEvent

davidChevrier committed:
Looking into PVs for triggering of Ampteks.

davidChevrier committed:
Pulled down the latest in Issue23 branch. Had to make some changes to get the dwell-trigger PVs working. A section in AmptekSDD123Server had accidentally been removed that was causing the lastRequestSpectrumTime_ and requestSpectrumTime_ to not be set properly. There was also an issue with the AMDSCentralServerSGM using the bufferGroupManagers_ where it should have been using dwellBufferGroupManagers_. Also needed to metaRegister the AMDSDwellStatusData type. Other minor changes.

XuetaoLiu-S committed:
Remove unused function `at` for AMDSBufferGroup

davidChevrier committed:
refine the project files for acquaman path

XuetaoLiu-S committed:
Merge Issue23

XuetaoLiu-S committed:
Some small code refactors based on David's changes

XuetaoLiu-S committed:
Merge pull request #79 from acquaman/Issue1679_Acquaman

PullRequest for Amptek Dwell request
